tutorials/haskell/haskell-time-tutorial/tutorial.md

```diff
@@ -0,0 +1,481 @@
+---
+title: Time, patterns and a little more
+published: 2021-11-25
+tags: haskell, pattern-synonyms, property-testing
+libraries: time
+language: haskell
+author: Felix Miño
+author-name: Felix Miño
+github-profile: felixminom
+description: In this tutorial we’re going to explore the use of pattern synonyms in Haskell, later we’ll give a brief overview to the Stack Builders’ contribution to the `time` Haskell library and finally we’ll review how to use the utilities that were introduced in the contribution.
+---
+
+As Stack Builders one of the core values of our company is contributing
+to open source (OSS). We believe that we have the power to change
+people’s lives (or at least make it easier) by pushing the boundaries
+of the software industry and we’re happy to announce that the past month
+we were able to contribute to one of Haskell's core libraries, time.
+
+One of our tech leads, Sebastian Estrella, started with the idea:
+“What if we bring some utilities that Ruby on Rails (RoR) has to Haskell?”.
+For instance, “How can I get the range of days that a quarter has?”
+or “How can I get a month’s boundaries?” were some of the questions
+that came to our mind. In the following example I will show the Haskell
+code needed to get a month's boundaries and its version in RoR.
+
+```haskell
+type MonthFirstDay = Day
+type MonthLastDay  = Day
+
+monthBoundaries :: Day -> (MonthFirstDay, MonthLastDay)
+monthBoundaries day = let (y, m , _) = toGregorian day
+   in (fromGregorian y m 1, fromGregorian y m 31)
+```
+
+```ruby
+today = Date.today
+today.next_month.all_month
+=> Wed, 01 Dec 2021..Fri, 31 Dec 2021
```

Comment:
I think this is still missing something to make the Haskell and Ruby codes comparable. The Haskell one shows the implementation (source), but the Ruby one shows how to use the implementation.

Assuming you have the Haskell implementation (which is the one in the example), then the Haskell equivalent would be something like this:

```
> today <- utctDay <$> getCurrentTime
> monthBoundaries today
(2021-11-01,2021-11-30)
```

Which is not that different compared to the Ruby one (ignoring that I skipped over the next month part, but the important thing is that the things being compared should match).

I think that the issue that you're trying to highlight is that with Haskell you have to implement the code because it's not part of the `time` library.

Reply:
This makes a lot of sense in that case I will rename the function. The RoR version actually returns a range so I will name the Haskell function `monthAllDays` and I will make it return a whole month (I was returning just starting and ending days). Also latter on, we can compare this with the `periodAllDays` that was implemented in the library, wdyt ?